instrumentation/mongo/mongo-common/src/main/java/io/opentelemetry/javaagent/instrumentation/mongo/MongoClientTracer.java

```diff
@@ -80,54 +97,111 @@ protected String dbConnectionString(CommandStartedEvent event) {
-    // scrub the Mongo command so that parameters are removed from the string
-    BsonDocument scrubbed = scrub(statement);
-    return scrubbed.toString();
+  private static final Method IS_TRUNCATED_METHOD;
+
+  static {
+    IS_TRUNCATED_METHOD = Arrays.stream(JsonWriter.class.getMethods())
+        .filter(method -> method.getName().equals("isTruncated")).findFirst().orElse(null);
   }
 
-  /**
-   * The values of these mongo fields will not be scrubbed out. This allows the non-sensitive
-   * collection names to be captured.
-   */
-  private static final List<String> UNSCRUBBED_FIELDS =
-      Arrays.asList("ordered", "insert", "count", "find", "create");
+  private JsonWriterSettings createJsonWriterSettings(int maxNormalizedQueryLength)  {
+    JsonWriterSettings settings = new JsonWriterSettings(false);
+    try {
+      // The static JsonWriterSettings.builder() method was introduced in the 3.5 release
+      Optional<Method> buildMethod = Arrays.stream(JsonWriterSettings.class.getMethods())
+          .filter(method -> method.getName().equals("builder")).findFirst();
+      if (buildMethod.isPresent()) {
+        Class<?> builderClass = buildMethod.get().getReturnType();
+        Object builder = buildMethod.get().invoke(null, (Object[]) null);
+
+        // The JsonWriterSettings.Builder.indent method was introduced in the 3.5 release,
+        // but checking anyway
+        Optional<Method> indentMethod = Arrays.stream(builderClass.getMethods())
+            .filter(method -> method.getName().equals("indent"))
+            .findFirst();
+        if (indentMethod.isPresent()) {
+          indentMethod.get().invoke(builder, false);
+        }
 
-  private static final BsonValue HIDDEN_CHAR = new BsonString("?");
+        // The JsonWriterSettings.Builder.maxLength method was introduced in the 3.7 release
+        Optional<Method> maxLengthMethod = Arrays.stream(builderClass.getMethods())
+            .filter(method -> method.getName().equals("maxLength"))
+            .findFirst();
+        if (maxLengthMethod.isPresent()) {
+          maxLengthMethod.get().invoke(builder, maxNormalizedQueryLength);
+        }
+        settings = (JsonWriterSettings) builderClass.getMethod("build", (Class<?>[]) null)
+            .invoke(builder, (Object[]) null);
+      }
+    } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException ignored) {
+    }
+    return settings;
+  }
 
-  private static BsonDocument scrub(BsonDocument origin) {
-    BsonDocument scrub = new BsonDocument();
+  @Override
+  public String normalizeQuery(BsonDocument command) {
+    StringWriter stringWriter = new StringWriter(128);
+    writeScrubbed(command, new JsonWriter(stringWriter, jsonWriterSettings), true);
+    // If using MongoDB driver >= 3.7, the substring invocation will be a no-op due to use of
+    // JsonWriterSettings.Builder.maxLength in the static initializer for JSON_WRITER_SETTINGS
+    return stringWriter.getBuffer()
```

Comment:
Question for reviewers: should we append a "..." to the query or just cut it off?  Or do we need to make sure that even when truncated it remains valid JSON?